File: ts/src/printer.ts
// printer.ts — port of Expression/Printer.swift
//
// Canonical AST -> string projection. Deterministic, minimally parenthesized.
//   - parse(print(ast)) ≡ ast
//   - print(parse(text)) is idempotent

import type { BinaryOperator, Expr, PathSegment } from "./ast.ts";
import { type JinjaValue, renderDouble, renderedString } from "./value.ts";

// Precedence ladder (higher binds tighter). Mirrors the parser.
const Prec = {
	ternary: 1,
	or: 2,
	and: 3,
	not: 4,
	comparison: 5,
	concat: 6,
	additive: 7,
	multiplicative: 8,
	negate: 9,
	filter: 10,
	reference: 11,
	primary: 12,
} as const;

function precedenceOf(op: BinaryOperator): number {
	switch (op) {
		case "==":
		case "!=":
		case "<":
		case "<=":
		case ">":
		case ">=":
		case "in":
		case "not in":
			return Prec.comparison;
		case "~":
			return Prec.concat;
		case "+":
		case "-":
			return Prec.additive;
		case "*":
		case "/":
		case "//":
		case "%":
			return Prec.multiplicative;
	}
}

/** Render an expression to its canonical source form. */
export function printExpr(expr: Expr): string {
	return emit(expr, 0);
}

function emit(expr: Expr, minContext: number): string {
	const [text, prec] = render(expr);
	return prec < minContext ? `(${text})` : text;
}

function render(expr: Expr): [string, number] {
	switch (expr.e) {
		case "literal":
			return [renderLiteral(expr.value), Prec.primary];
		case "reference":
			return [renderReference(expr.segments), Prec.reference];
		case "unary":
			if (expr.op === "not") {
				return [`not ${emit(expr.operand, Prec.not)}`, Prec.not];
			}
			return [`-${emit(expr.operand, Prec.negate)}`, Prec.negate];
		case "binary": {
			const p = precedenceOf(expr.op);
			const left = emit(expr.lhs, p);
			const right = emit(expr.rhs, p + 1);
			return [`${left} ${expr.op} ${right}`, p];
		}
		case "logical": {
			const p = expr.op === "or" ? Prec.or : Prec.and;
			const keyword = expr.op === "or" ? "or" : "and";
			const parts = expr.operands.map((o) => emit(o, p + 1));
			return [parts.join(` ${keyword} `), p];
		}
		case "conditional": {
			const thenText = emit(expr.then, Prec.or);
			const condText = emit(expr.condition, Prec.or);
			const elseText = emit(expr.otherwise, Prec.ternary);
			return [`${thenText} if ${condText} else ${elseText}`, Prec.ternary];
		}
		case "filter": {
			const inputText = emit(expr.input, Prec.filter);
			if (expr.arguments.length === 0) {
				return [`${inputText} | ${expr.name}`, Prec.filter];
			}
			const argText = expr.arguments.map((a) => emit(a, 0)).join(", ");
			return [`${inputText} | ${expr.name}(${argText})`, Prec.filter];
		}
		case "arrayLiteral": {
			const parts = expr.elements.map((el) => emit(el, 0));
			return [`[${parts.join(", ")}]`, Prec.primary];
		}
		case "objectLiteral": {
			const parts = expr.entries.map((en) => `${renderKey(en.key)}: ${emit(en.value, 0)}`);
			return [`{${parts.join(", ")}}`, Prec.primary];
		}
	}
}

// MARK: - Leaves

function renderLiteral(value: JinjaValue): string {
	switch (value.kind) {
		case "null":
			return "null";
		case "bool":
			return value.value ? "true" : "false";
		case "int":
			return String(value.value);
		case "double":
			return renderDouble(value.value);
		case "string":
			return renderStringLiteral(value.value);
		case "date":
			// No source syntax for dates; best-effort quoted ISO string.
			return renderStringLiteral(renderedString(value));
		case "array": {
			const parts = value.value.map(renderLiteral);
			return `[${parts.join(", ")}]`;
		}
		case "object": {
			// Sort by key so a literal object prints canonically.
			const parts = [...value.value.entries()]
				.sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
				.map(([k, v]) => `${renderKey(k)}: ${renderLiteral(v)}`);
			return `{${parts.join(", ")}}`;
		}
	}
}

function renderStringLiteral(s: string): string {
	let out = "'";
	for (const ch of s) {
		switch (ch) {
			case "\\":
				out += "\\\\";
				break;
			case "'":
				out += "\\'";
				break;
			case "\n":
				out += "\\n";
				break;
			case "\t":
				out += "\\t";
				break;
			case "\r":
				out += "\\r";
				break;
			default:
				out += ch;
		}
	}
	out += "'";
	return out;
}

function renderReference(segments: PathSegment[]): string {
	let out = "";
	segments.forEach((segment, i) => {
		switch (segment.s) {
			case "key":
				if (i === 0) {
					out += segment.name;
				} else if (isIdentifier(segment.name)) {
					out += `.${segment.name}`;
				} else {
					out += `[${renderStringLiteral(segment.name)}]`;
				}
				break;
			case "index":
				out += `[${segment.index}]`;
				break;
			case "dynamic":
				out += `[${emit(segment.expr, 0)}]`;
				break;
		}
	});
	return out;
}

function renderKey(key: string): string {
	return isIdentifier(key) ? key : renderStringLiteral(key);
}

const KEYWORDS = new Set([
	"true",
	"false",
	"null",
	"none",
	"and",
	"or",
	"not",
	"in",
	"if",
	"else",
]);

function isIdentifier(s: string): boolean {
	const chars = Array.from(s);
	const first = chars[0];
	if (first === undefined || !(/\p{L}/u.test(first) || first === "_")) return false;
	for (let i = 1; i < chars.length; i++) {
		const ch = chars[i]!;
		if (!(/\p{L}/u.test(ch) || /\p{Nd}/u.test(ch) || ch === "_")) return false;
	}
	return !KEYWORDS.has(s);
}

File: Sources/CococoJinja/Expression/Printer.swift
//
//  Printer.swift
//  CococoJinja
//
//  Copyright Triple C Labs GmbH
//
//  Canonical AST -> string projection. Deterministic, minimally parenthesized.
//  Guarantees (see docs/workflow-expression-dialect.md §5):
//    - parse(print(ast)) ≡ ast        (parens are transparent in the AST)
//    - print(parse(text)) is idempotent (canonical normalization)
//
//  This prints the *expression* form (no surrounding `{{ }}`); the template
//  layer adds those when rendering a stored condition string.
//

import Foundation

enum Printer {
    /// Render an expression to its canonical source form.
    static func print(_ expr: Expr) -> String {
        emit(expr, minContext: 0)
    }

    // MARK: - Precedence

    // Higher binds tighter. Mirrors Parser's ladder.
    private enum Prec {
        static let ternary = 1
        static let or = 2
        static let and = 3
        static let not = 4
        static let comparison = 5
        static let concat = 6
        static let additive = 7
        static let multiplicative = 8
        static let negate = 9
        static let filter = 10
        static let reference = 11
        static let primary = 12
    }

    private static func precedence(of op: BinaryOperator) -> Int {
        switch op {
        case .eq, .neq, .lt, .lte, .gt, .gte, .contains, .notContains: Prec.comparison
        case .concat: Prec.concat
        case .add, .subtract: Prec.additive
        case .multiply, .divide, .floorDivide, .modulo: Prec.multiplicative
        }
    }

    /// Render `expr`, wrapping in parentheses if its own precedence is below the
    /// surrounding `minContext`.
    private static func emit(_ expr: Expr, minContext: Int) -> String {
        let (text, prec) = render(expr)
        return prec < minContext ? "(\(text))" : text
    }

    /// Returns the rendered text and the node's precedence.
    private static func render(_ expr: Expr) -> (String, Int) {
        switch expr {
        case let .literal(value):
            return (renderLiteral(value), Prec.primary)

        case let .reference(segments):
            return (renderReference(segments), Prec.reference)

        case let .unary(op, operand):
            switch op {
            case .not:
                return ("not " + emit(operand, minContext: Prec.not), Prec.not)
            case .negate:
                return ("-" + emit(operand, minContext: Prec.negate), Prec.negate)
            }

        case let .binary(op, lhs, rhs):
            let p = precedence(of: op)
            // Left-associative: left child allowed at this precedence, right child
            // must bind strictly tighter (else it needs parens).
            let left = emit(lhs, minContext: p)
            let right = emit(rhs, minContext: p + 1)
            return ("\(left) \(op.rawValue) \(right)", p)

        case let .logical(op, operands):
            let p = op == .or ? Prec.or : Prec.and
            let keyword = op == .or ? "or" : "and"
            // Operands bind strictly tighter than this level.
            let parts = operands.map { emit($0, minContext: p + 1) }
            return (parts.joined(separator: " \(keyword) "), p)

        case let .conditional(condition, then, otherwise):
            // `then if condition else otherwise` — then/condition at or-level,
            // otherwise stays at ternary level (right-associative chaining).
            let thenText = emit(then, minContext: Prec.or)
            let condText = emit(condition, minContext: Prec.or)
            let elseText = emit(otherwise, minContext: Prec.ternary)
            return ("\(thenText) if \(condText) else \(elseText)", Prec.ternary)

        case let .filter(name, input, arguments):
            let inputText = emit(input, minContext: Prec.filter)
            if arguments.isEmpty {
                return ("\(inputText) | \(name)", Prec.filter)
            }
            let argText = arguments.map { emit($0, minContext: 0) }.joined(separator: ", ")
            return ("\(inputText) | \(name)(\(argText))", Prec.filter)

        case let .arrayLiteral(elements):
            let parts = elements.map { emit($0, minContext: 0) }
            return ("[\(parts.joined(separator: ", "))]", Prec.primary)

        case let .objectLiteral(entries):
            let parts = entries.map { "\(renderKey($0.key)): \(emit($0.value, minContext: 0))" }
            return ("{\(parts.joined(separator: ", "))}", Prec.primary)
        }
    }

    // MARK: - Leaves

    private static func renderLiteral(_ value: JinjaValue) -> String {
        switch value {
        case .null: return "null"
        case let .bool(b): return b ? "true" : "false"
        case let .int(i): return String(i)
        case let .double(d): return String(d)
        case let .string(s): return renderStringLiteral(s)
        case let .date(d):
            // Dates have no source literal syntax; the parser never produces one.
            // Best-effort for programmatically-built ASTs (not round-trippable).
            return renderStringLiteral(JinjaValue.date(d).renderedString())
        case let .array(items):
            let parts = items.map { renderLiteral($0) }
            return "[\(parts.joined(separator: ", "))]"
        case let .object(dict):
            // Sort by key so a literal object prints canonically regardless of the
            // backing dictionary's iteration order.
            let parts = dict.sorted { $0.key < $1.key }
                .map { "\(renderKey($0.key)): \(renderLiteral($0.value))" }
            return "{\(parts.joined(separator: ", "))}"
        }
    }

    private static func renderStringLiteral(_ s: String) -> String {
        var out = "'"
        for ch in s {
            switch ch {
            case "\\": out += "\\\\"
            case "'": out += "\\'"
            case "\n": out += "\\n"
            case "\t": out += "\\t"
            case "\r": out += "\\r"
            default: out.append(ch)
            }
        }
        out += "'"
        return out
    }

    private static func renderReference(_ segments: [PathSegment]) -> String {
        var out = ""
        for (i, segment) in segments.enumerated() {
            switch segment {
            case let .key(name):
                if i == 0 {
                    // Root is always a bare identifier (the parser only makes
                    // references from an identifier root).
                    out += name
                } else if isIdentifier(name) {
                    out += ".\(name)"
                } else {
                    out += "[\(renderStringLiteral(name))]"
                }
            case let .index(n):
                out += "[\(n)]"
            case let .dynamic(expr):
                out += "[\(emit(expr, minContext: 0))]"
            }
        }
        return out
    }

    /// An object/member key prints bare when it's a valid identifier, else quoted.
    private static func renderKey(_ key: String) -> String {
        isIdentifier(key) ? key : renderStringLiteral(key)
    }

    /// Whether `s` lexes as a single identifier token (and isn't a keyword).
    private static func isIdentifier(_ s: String) -> Bool {
        guard let first = s.first, first.isLetter || first == "_" else { return false }
        for ch in s.dropFirst() where !(ch.isLetter || ch.isNumber || ch == "_") {
            return false
        }
        switch s {
        case "true", "false", "null", "none", "and", "or", "not", "in", "if", "else":
            return false
        default:
            return true
        }
    }
}
